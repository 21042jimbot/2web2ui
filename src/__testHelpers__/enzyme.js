export const findButton = (wrapper, buttonLabel) =>
  wrapper.findWhere(
    (node) => node.type() === 'button' && new RegExp(buttonLabel).test(node.text())
  );
