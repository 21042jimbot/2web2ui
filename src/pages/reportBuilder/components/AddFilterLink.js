import React, { useState } from 'react';
import qs from 'qs';
import { useLocation } from 'react-router-dom';
import { ButtonLink } from 'src/components/links';
import { ScreenReaderOnly } from 'src/components/matchbox';
import { useReportBuilderContext } from '../context/ReportBuilderContext';
import { stringifyTypeaheadfilter } from 'src/helpers/string';

export const AddFilterLink = ({ newFilter }) => {
  const [isNewTabKeyPressed, setNewTabKeyPressed] = useState(false);
  const { actions, selectors } = useReportBuilderContext();
  const { addFilters } = actions;
  const { selectSummaryChartSearchOptions: currentSearchOptions } = selectors;
  const { pathname } = useLocation();

  /**
   * Needs to first check if cmd/ctrl key is pressed when mouse is released
   * Then it's saved for use in the onClick handler. We need both because
   * onClick does not have the correct values for metaKey and ctrlKey
   **/
  const handleMouseUp = e => {
    setNewTabKeyPressed(e.metaKey || e.ctrlKey);
  };

  const handleClick = e => {
    if (!isNewTabKeyPressed) {
      addFilters(newFilter);
      // eslint-disable-next-line no-unused-expressions
      e?.preventDefault();
    }
  };

  const currentFilters = currentSearchOptions.filters || [];
  const mergedFilters = [...currentFilters, stringifyTypeaheadfilter(newFilter)];
  const newSearchOptions = { ...currentSearchOptions, filters: mergedFilters };

  const linkParams = qs.stringify(newSearchOptions);
  const fullLink = `${pathname}?${linkParams}`;

  return (
    <ButtonLink onMouseUp={handleMouseUp} onClick={handleClick} to={fullLink}>
      {newFilter.value}

      <ScreenReaderOnly>(Applies a filter to the report)</ScreenReaderOnly>
    </ButtonLink>
  );
};

export default AddFilterLink;
