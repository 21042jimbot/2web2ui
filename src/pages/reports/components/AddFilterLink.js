import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import qs from 'qs';

import { addFilters } from 'src/actions/reportOptions';
import { stringifyTypeaheadfilter } from 'src/helpers/string';
import { selectReportSearchOptions, selectSummaryChartSearchOptions } from 'src/selectors/reportSearchOptions';
import { UnstyledLink } from '@sparkpost/matchbox';

export const AddFilterLink = ({
  //From parent
  newFilter,
  reportType,
  content,
  //From redux
  summarySearchOptions,
  reportSearchOptions,
  addFilters }) => {

  const searchOptions = reportType === 'summary' ? summarySearchOptions : reportSearchOptions; //summarySearchOptions includes selected metrics options
  const oldFilters = searchOptions.filters || [];

  const handleOnClick = (e) => {
    addFilters([newFilter]);
    //This prevents the browser from following the link and reloading the page.
    e.preventDefault();
  };

  const mergedFilters = _.uniqWith([ ...oldFilters, stringifyTypeaheadfilter(newFilter)], _.isEqual);
  const linkParams = qs.stringify({ ...searchOptions, filters: mergedFilters }, { arrayFormat: 'repeat' });
  const fullLink = `/reports/${reportType}/?${linkParams}`;

  return (
    <UnstyledLink onClick={handleOnClick} to={fullLink}>{content}</UnstyledLink>
  );
};

const mapStateToProps = (state) => ({
  summarySearchOptions: selectSummaryChartSearchOptions(state),
  reportSearchOptions: selectReportSearchOptions(state)
});

export default connect(mapStateToProps, { addFilters })(AddFilterLink);
