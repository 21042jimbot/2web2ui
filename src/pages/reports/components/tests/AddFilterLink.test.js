import React from 'react';
import { AddFilterLink } from '../AddFilterLink';
import { shallow } from 'enzyme';

describe('Add Filter Link', () => {

  const filters = ['Campaign: shiny new filter'];
  const baseProps = {
    newFilter: { id: 0, type: 'Subaccount', value: 'Master Account (ID 0)' },
    reportType: 'summary',
    content: 'master account',
    addFilters: jest.fn(),
    reportSearchOptions: {
      filters
    },
    summarySearchOptions: {
      filters,
      metrics: ['count_something']
    }
  };

  function subject(props) {
    return shallow(<AddFilterLink {...baseProps} {...props} />);
  }

  it('should render correctly for the summary report and include metrics in the link', () => {
    const wrapper = subject();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly for other reports and should not include metrics in the link', () => {
    const wrapper = subject({ reportType: 'bounce' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle click correctly', () => {
    const wrapper = subject();
    const preventDefaultMock = jest.fn();
    wrapper.find('UnstyledLink').simulate('click', { preventDefault: preventDefaultMock });

    expect(baseProps.addFilters).toHaveBeenCalledWith([{ id: 0, type: 'Subaccount', value: 'Master Account (ID 0)' }]);
    expect(preventDefaultMock).toHaveBeenCalled();
  });
});
