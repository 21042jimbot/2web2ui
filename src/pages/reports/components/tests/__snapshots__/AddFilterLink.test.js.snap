// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Filter Link should render correctly for other reports and should not include metrics in the link 1`] = `
<UnstyledLink
  onClick={[Function]}
  to="/reports/bounce/?filters=Campaign%3A%20shiny%20new%20filter&filters=Subaccount%3AMaster%20Account%20%28ID%200%29%3A0"
>
  master account
</UnstyledLink>
`;

exports[`Add Filter Link should render correctly for the summary report and include metrics in the link 1`] = `
<UnstyledLink
  onClick={[Function]}
  to="/reports/summary/?filters=Campaign%3A%20shiny%20new%20filter&filters=Subaccount%3AMaster%20Account%20%28ID%200%29%3A0&metrics=count_something"
>
  master account
</UnstyledLink>
`;
